OpenRA.Mods.Common/Traits/Transforms.cs

```diff
@@ -115,23 +117,31 @@ public void DeployTransform(bool queued)
 			if (self.Info.HasTraitInfo<AircraftInfo>())
 				self.QueueActivity(new HeliLand(self, true));
 
-			foreach (var nt in self.TraitsImplementing<INotifyTransform>())
-				nt.BeforeTransform(self);
-
-			var transform = new Transform(self, info.IntoActor)
+			self.QueueActivity(new CallFunc(() =>
 			{
-				Offset = info.Offset,
-				Facing = info.Facing,
-				Sounds = info.TransformSounds,
-				Notification = info.TransformNotification,
-				Faction = faction
-			};
-
-			var makeAnimation = self.TraitOrDefault<WithMakeAnimation>();
-			if (makeAnimation != null)
-				makeAnimation.Reverse(self, transform);
-			else
-				self.QueueActivity(transform);
+				// Prevent deployment in bogus locations
+				var building = self.TraitOrDefault<Building>();
+				if (!CanDeploy() || (building != null && !building.Lock()))
+					return;
+
+				foreach (var nt in self.TraitsImplementing<INotifyTransform>())
+					nt.BeforeTransform(self);
+
+				var transform = new Transform(self, info.IntoActor)
+				{
+					Offset = info.Offset,
+					Facing = info.Facing,
+					Sounds = info.TransformSounds,
+					Notification = info.TransformNotification,
+					Faction = faction
+				};
+
+				var makeAnimation = self.TraitOrDefault<WithMakeAnimation>();
```

Comment:
I think you should add a `self.CancelActivity()` here, since otherwise the player might inject other orders between the `CallFunc` and the `Transform`. It's quite easy to reproduce with some move orders. The mcv would change facing at some totally inappropriate time and then run the make animation without turning into the correct direction.